src/main/java/io/vertx/core/spi/cluster/impl/DefaultDeliveryStrategy.java

```diff
@@ -0,0 +1,197 @@
+
+  private final ConcurrentMap<String, NodeSelector> selectors = new ConcurrentHashMap<>();
+
+  private ClusterManager clusterManager;
+  private String nodeId;
+
+  @Override
+  public void setVertx(VertxInternal vertx) {
+    clusterManager = vertx.getClusterManager();
+  }
+
+  @Override
+  public void eventBusStarted() {
+    nodeId = clusterManager.getNodeId();
+  }
+
+  @Override
+  public Future<List<String>> chooseNodes(Message<?> message) {
+    return selector(message.address()).map(selector -> selector.selectNodes(message));
+  }
+
+  public Future<NodeSelector> selector(String address) {
+    ContextInternal context = (ContextInternal) Vertx.currentContext();
+    Objects.requireNonNull(context, "Method must be invoked on a Vert.x thread");
+
+    Queue<Promise<NodeSelector>> waiters = getWaiters(context, address);
+
+    NodeSelector selector = selectors.get(address);
+    if (selector != null && waiters.isEmpty()) {
+      return context.succeededFuture(selector);
+    }
+
+    Promise<NodeSelector> promise = context.promise();
+
+    waiters.add(promise);
+    if (waiters.size() == 1) {
+      dequeueWaiters(context, address);
+    }
+    return promise.future();
+  }
+
+  @SuppressWarnings("unchecked")
+  private Queue<Promise<NodeSelector>> getWaiters(ContextInternal context, String address) {
+    Map<String, Queue<Promise<NodeSelector>>> map = (Map<String, Queue<Promise<NodeSelector>>>) context.contextData().computeIfAbsent(this, ctx -> new HashMap<>());
+    return map.computeIfAbsent(address, a -> new ArrayDeque<>());
+  }
+
+  @SuppressWarnings("unchecked")
+  private void removeWaiters(ContextInternal context, String address) {
+    context.contextData().compute(this, (k, v) -> {
+      Map<String, Queue<Promise<NodeSelector>>> map = (Map<String, Queue<Promise<NodeSelector>>>) v;
+      if (map == null) {
+        throw new IllegalStateException();
+      }
+      map.remove(address);
+      return !map.isEmpty() ? map : null;
+    });
+  }
+
+  private void dequeueWaiters(ContextInternal context, String address) {
+    if (Vertx.currentContext() != context) {
+      throw new IllegalStateException();
+    }
+
+    Queue<Promise<NodeSelector>> waiters = getWaiters(context, address);
+
+    Promise<NodeSelector> waiter;
+    for (int i = 0; ; i++) {
+      if (i == 10) {
+        // Give a chance to other tasks every 10 iterations
+        context.runOnContext(v -> dequeueWaiters(context, address));
+      }
+      Promise<NodeSelector> peeked = waiters.peek();
+      if (peeked == null) {
+        throw new IllegalStateException();
+      }
+      NodeSelector selector = selectors.get(address);
+      if (selector != null) {
+        peeked.complete(selector);
+        waiters.remove();
+        if (waiters.isEmpty()) {
+          removeWaiters(context, address);
+          return;
+        }
+      } else {
+        waiter = peeked;
+        break;
+      }
+    }
+
+    clusterManager.registrationListener(address)
```

Comment:
instead use a single `onComplete(...)` operation, using `onFailure`/`onSuccess` leads to create un-necessary objects (it will create in practice two `onComplete` listeners)

Reply:
Done